utils/dataset_manifest/utils.py

```diff
@@ -21,4 +24,112 @@ def rotate_image(image, angle):
+        ]
+        context_dirs = (os.path.join(dirname_before_before, match.group(), 'data') for match in matches if match)
+        potentially_related_images = (
+            os.path.join(context_dir, context_image) for context_dir in context_dirs for context_image in os.listdir(context_dir)
+        )
+        potentially_related_images = filter(is_image, potentially_related_images)
+        related_images = filter(lambda x: os.path.splitext(os.path.basename(x))[0] == name, potentially_related_images)
+        related_images = map(lambda x: os.path.relpath(x, root_path), related_images)
+        return relative_image_path, sorted(list(related_images))
+
+    # 3D POINTCLOUD DATA FORMAT
+    # pointcloud/
+    #     00001.pcd
+    # related_images/
+    #     00001_pcd/
+    #         image_01.png # or other image
+
+    # 2D DATFORMAT
+    # data/
+    #     00001.png
+    #     related_images/
+    #         00001_png/
+    #             context_image_1.jpeg
+    #             context_image_2.png
+    elif dirname_related_images_exist or dirname_before_related_image_exist:
+        converted_name = '_'.join((name, ext[1:])) if ext else name
+        related_images_dir = os.path.join(dirname if dirname_related_images_exist else dirname_before, 'related_images', converted_name)
+        if os.path.isdir(related_images_dir):
+            relative_related_path = os.path.relpath(related_images_dir, root_path)
+            return relative_image_path, sorted(
+                list(
+                    map(
+                        lambda x: os.path.join(relative_related_path, x),
+                        filter(lambda x: is_image(os.path.join(related_images_dir, x)), os.listdir(related_images_dir))
+                    )
+                )
+            )
+
+    # 3D, DEFAULT DATAFORMAT
+    # Option 1
+    # data/
+    #     image.pcd
+    #     image.png
+
+    # Option 2
+    # data/
+    #    image_1/
+    #        image_1.pcd
+    #        context_1.png
+    #        context_2.jpg
+    else:
+        base_path = os.path.relpath(dirname, root_path)
+        all_files = filter(lambda x: x != ''.join((name, ext)), os.listdir(dirname))
+
+        if os.path.basename(dirname) != name: # Option 1
+            all_files = filter(lambda x: os.path.splitext(x)[0] == name, all_files)
+
+        return relative_image_path, sorted(
+            list(
+                map(
+                    lambda x: os.path.join(base_path, x),
+                    filter(lambda x: is_image(os.path.join(dirname, x)), all_files)
+                )
+            )
+        )
+
+    return relative_image_path, []
+
+
+# image_path is expected to be a list of absolute path to images
+# root_path is expected to be a string (dataset root)
+def detect_related_images(image_paths, root_path):
+    related_images = {k: v for k, v in map(lambda path: detect_related_images_for(path, root_path), image_paths)}
```

Comment:
@bsekachev , right now detect_related_images_for will call a couple of syscalls for every image. I believe it can be a problem for a big list of images. What is a reason to look for "related_images" directory using the way?

Reply:
Agree, should be optimized

Reply:
The approach was redesigned